presto-main/src/main/java/com/facebook/presto/operator/aggregation/multimapagg/MultimapAggregationFunction.java

```diff
@@ -109,44 +115,47 @@ private static InternalAggregationFunction generateAggregation(Type keyType, Typ
                 new ParameterMetadata(BLOCK_INDEX));
     }
 
-    public static void input(MultiKeyValuePairsState state, Block key, Block value, int position)
+    public static void input(MultimapAggregationState state, Block key, Block value, int position)
     {
-        MultiKeyValuePairs pairs = state.get();
-        if (pairs == null) {
-            pairs = new MultiKeyValuePairs(state.getKeyType(), state.getValueType());
-            state.set(pairs);
-        }
-
-        long startSize = pairs.estimatedInMemorySize();
-        pairs.add(key, value, position, position);
-        state.addMemoryUsage(pairs.estimatedInMemorySize() - startSize);
+        state.add(key, value, position);
     }
 
-    public static void combine(MultiKeyValuePairsState state, MultiKeyValuePairsState otherState)
+    public static void combine(MultimapAggregationState state, MultimapAggregationState otherState)
     {
-        if (state.get() != null && otherState.get() != null) {
-            Block keys = otherState.get().getKeys();
-            Block values = otherState.get().getValues();
-            MultiKeyValuePairs pairs = state.get();
-            long startSize = pairs.estimatedInMemorySize();
-            for (int i = 0; i < keys.getPositionCount(); i++) {
-                pairs.add(keys, values, i, i);
-            }
-            state.addMemoryUsage(pairs.estimatedInMemorySize() - startSize);
-        }
-        else if (state.get() == null) {
-            state.set(otherState.get());
-        }
+        state.merge(otherState);
     }
 
-    public static void output(MultiKeyValuePairsState state, BlockBuilder out)
+    public static void output(Type keyType, Type valueType, MultimapAggregationState state, BlockBuilder out)
     {
-        MultiKeyValuePairs pairs = state.get();
-        if (pairs == null) {
+        if (state.isEmpty()) {
             out.appendNull();
         }
         else {
-            pairs.toMultimapNativeEncoding(out);
+            // TODO: this behavior is adopted from the now defunct MultiKeyValuePairs.java
+            // Investigate utilizing the behavior found in multimap_from_entries which avoids rewriting the values in valueArrayBlockBuilders
+            ObjectBigArray<BlockBuilder> valueArrayBlockBuilders = new ObjectBigArray<>();
+            BlockBuilder distinctKeyBlockBuilder = keyType.createBlockBuilder(null, state.getEntryCount(), expectedValueSize(keyType, 100));
+            TypedSet keySet = new TypedSet(keyType, state.getEntryCount(), MultimapAggregationFunction.NAME);
+
+            state.forEach((key, value, keyValueIndex) -> {
+                // If the the key is not present in the keyset, add it, and note its position in the keyset.  Its position in the
```

Comment:
I think we can use the old comment.

```
// Merge values of the same key into an array
```

In general I prefer to write comment stating the purpose instead of describe what the code did, which can be done by just reading the code.